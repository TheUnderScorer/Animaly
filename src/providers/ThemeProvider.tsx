import React, {
  createContext,
  FC,
  useCallback,
  useContext,
  useState,
} from 'react';
import { ApplicationProvider, IconRegistry } from '@ui-kitten/components';
import * as eva from '@eva-design/eva';
import { EvaIconsPack } from '@ui-kitten/eva-icons';

const defaultValue: ThemeContextType = {
  theme: 'light',
  toggleTheme: () => {},
};

export interface ThemeContextType {
  theme: 'light' | 'dark';
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType>(defaultValue);
export const useThemeContext = () => useContext(ThemeContext);

const ThemeProvider: FC = ({ children }) => {
  const [theme, setTheme] = useState<'light' | 'dark'>('light');
  const toggleTheme = useCallback(() => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  }, [theme]);

  return (
    <ThemeContext.Provider
      value={{
        theme,
        toggleTheme,
      }}>
      <IconRegistry icons={EvaIconsPack} />
      <ApplicationProvider {...eva} theme={eva[theme]}>
        {children}
      </ApplicationProvider>
    </ThemeContext.Provider>
  );
};

export default ThemeProvider;
