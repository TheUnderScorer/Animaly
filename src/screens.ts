export enum RootScreens {
  Home = 'Home',
  Splash = 'Splash',
}

export enum SplashScreens {
  Start = 'Start',
  CreateUserForm = 'CreateUserForm',
}

export enum AppScreens {
  Welcome = 'Welcome',
  Home = 'Home',
}
