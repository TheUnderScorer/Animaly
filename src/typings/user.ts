export interface User {
  name: string;
  createdAt: Date;
}
