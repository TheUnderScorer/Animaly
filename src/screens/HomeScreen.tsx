import React, { FC } from 'react';
import { useNavigation } from '@react-navigation/native';
import { CenteredLayout } from '../styles/view';
import { Button, Icon, Text } from '@ui-kitten/components';
import { RootScreens } from '../screens';
import { normalIcon } from '../styles/icons';

export interface HomeScreenProps {}

const HomeScreen: FC<HomeScreenProps> = () => {
  const navigation = useNavigation();

  return (
    <CenteredLayout level="1">
      <Text category="h1">Hello, React!</Text>
      <Button
        onPress={() => navigation.navigate(RootScreens.Splash)}
        appearance="filled"
        size="large"
        accessoryLeft={() => (
          <Icon fill="#8F9BB3" style={normalIcon} name="star" />
        )}>
        Click me!
      </Button>
    </CenteredLayout>
  );
};

export default HomeScreen;
