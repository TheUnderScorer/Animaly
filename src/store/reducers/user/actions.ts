import { Dispatch } from 'redux';
import { AppActions } from '../../index';
import AsyncStore from '../../../storage/AsyncStore';
import { User } from '../../../typings/user';
import { CreateUserInput } from './types';

export const fetchCurrentUser = (asyncStore: AsyncStore) => async (
  dispatch: Dispatch<AppActions>,
) => {
  const currentUser = await asyncStore.get<User>('currentUser');

  dispatch({
    type: 'SetCurrentUser',
    payload: currentUser,
  });

  return dispatch({
    type: 'SetDidInitialFetch',
    payload: true,
  });
};

export const createUser = (
  asyncStore: AsyncStore,
  input: CreateUserInput,
) => async (dispatch: Dispatch<AppActions>) => {
  const user: User = {
    ...input,
    createdAt: new Date(),
    hasBeenWelcomed: false,
  };

  asyncStore.set('currentUser', user);

  dispatch({
    type: 'SetCurrentUser',
    payload: user,
  });

  return user;
};
