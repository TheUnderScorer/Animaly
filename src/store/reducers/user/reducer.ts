import { User } from '../../../typings/user';
import { Action, defaultHandler, mapReducer } from '@theunderscorer/utils';

export interface UserState {
  currentUser: User | null;
  didInitialFetch: boolean;
}

export const initialState: UserState = {
  currentUser: null,
  didInitialFetch: false,
};

export type UserActions =
  | Action<'SetCurrentUser', User | null>
  | Action<'SetDidInitialFetch', boolean>;

export const userReducer = mapReducer<UserState, UserActions>(
  {
    SetCurrentUser: defaultHandler('currentUser'),
    SetDidInitialFetch: defaultHandler('didInitialFetch'),
  },
  initialState,
);
