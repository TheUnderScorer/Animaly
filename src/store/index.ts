import { applyMiddleware, combineReducers, createStore } from 'redux';
import { UserActions, userReducer, UserState } from './reducers/user/reducer';
import thunk from 'redux-thunk';
import { initialState as userInitialState } from './reducers/user/reducer';

export interface AppStore {
  user: UserState;
}

const initialState: AppStore = {
  user: userInitialState,
};

export type AppActions = UserActions;

const reducers = combineReducers({
  user: userReducer,
});
const middleware = [thunk];

const store = createStore(
  reducers,
  initialState,
  applyMiddleware(...middleware),
);

export default store;
