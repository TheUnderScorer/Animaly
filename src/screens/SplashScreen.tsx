import React, {
  FC,
  MutableRefObject,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Layout, Spinner } from '@ui-kitten/components';
import { RowView } from '../styles/view';
import {
  Image,
  StatusBar,
  StyleSheet,
  View,
  Text,
  Linking,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { AppStore } from '../store';
import {
  Transition,
  Transitioning,
  TransitioningView,
} from 'react-native-reanimated';
import { useAsyncStorageContext } from '../providers/AsyncStorageProvider';
import { fetchCurrentUser } from '../store/reducers/user/actions';
import AppLogo from '../ui/molecules/AppLogo';
import { useNavigation } from '@react-navigation/native';
import { RootScreens } from '../screens';
import { wait } from '../utils/timeout';
import CreateUserForm from '../ui/organisms/CreateUserForm';
import GetStartedBtn from '../ui/atoms/GetStartedBtn';
import Bg from '../assets/paws.jpg';
import { BackgroundOverlay } from '../styles/overlay';
import { A } from '../styles/typography';

export interface SplashScreenProps {}

const styles = StyleSheet.create({
  logo: {
    marginBottom: 40,
  },
  row: {
    justifyContent: 'center',
    alignItems: 'center',
    width: 300,
  },
  layout: {
    width: '100%',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  bg: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    zIndex: 0,
    top: 0,
  },
  bgSource: {
    position: 'absolute',
    color: '#fff',
    bottom: 20,
    left: 10,
  },
});

export type CreationState = 'notCreating' | 'creating';

const SplashScreen: FC<SplashScreenProps> = () => {
  const transitionRef = useRef<TransitioningView>();

  const [creationState, setCreationState] = useState<CreationState>(
    'notCreating',
  );

  const navigation = useNavigation();

  const { storage } = useAsyncStorageContext();

  const dispatch = useDispatch();
  const currentUser = useSelector<AppStore>((store) => store.user.currentUser);
  const didInitialUserFetch = useSelector<AppStore>(
    (store) => store.user.didInitialFetch,
  );
  const [loading, setLoading] = useState(false);

  const handleToggleIsCreatingUser = useCallback(() => {
    transitionRef.current!.animateNextTransition();
    setCreationState('creating');
  }, [transitionRef]);

  useEffect(() => {
    if (didInitialUserFetch || loading || !transitionRef.current) {
      return;
    }

    const timeout = setTimeout(async () => {
      transitionRef.current!.animateNextTransition();

      setLoading(true);

      await wait(1000);

      await dispatch(fetchCurrentUser(storage));

      setLoading(false);
    }, 1000);

    return () => {
      clearTimeout(timeout);
    };
  }, [
    didInitialUserFetch,
    currentUser,
    dispatch,
    loading,
    storage,
    transitionRef,
  ]);

  useEffect(() => {
    if (didInitialUserFetch) {
      if (currentUser) {
        navigation.navigate(RootScreens.Home);
      }
    }
  }, [didInitialUserFetch, currentUser, navigation, transitionRef]);

  return (
    <Transitioning.View
      style={styles.container}
      ref={transitionRef as MutableRefObject<TransitioningView>}
      transition={
        <Transition.Sequence>
          <Transition.Out type="fade" />
          <Transition.In type="fade" delayMs={500} />
          <Transition.Change interpolation="easeInOut" />
        </Transition.Sequence>
      }>
      <Layout level="4" style={styles.layout}>
        {creationState === 'notCreating' && (
          <>
            <Image source={Bg} style={styles.bg} />
            <BackgroundOverlay />
            <Text style={styles.bgSource}>
              Designed by{' '}
              <A
                variant="dark"
                onPress={() =>
                  Linking.openURL('https://www.freepik.com/pch-vector')
                }>
                pch-vector
              </A>
            </Text>
          </>
        )}
        <StatusBar hidden />
        {creationState === 'creating' && <CreateUserForm />}
        <View>
          {creationState === 'notCreating' && (
            <View style={styles.logo}>
              <AppLogo />
            </View>
          )}
          <RowView style={[styles.row]}>
            {loading && <Spinner testID="splashSpinner" size="giant" />}
            {didInitialUserFetch &&
              !currentUser &&
              creationState === 'notCreating' && (
                <GetStartedBtn onPress={handleToggleIsCreatingUser} />
              )}
          </RowView>
        </View>
      </Layout>
    </Transitioning.View>
  );
};

export default SplashScreen;
