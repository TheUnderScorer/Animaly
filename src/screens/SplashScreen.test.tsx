import React, { ReactNode } from 'react';
import store from '../store';
import AsyncStorageProvider from '../providers/AsyncStorageProvider';
import ThemeProvider from '../providers/ThemeProvider';
import { NavigationContainer } from '@react-navigation/native';
import { Provider } from 'react-redux';
import { act, render } from 'react-native-testing-library';
import SplashScreen from './SplashScreen';
import { wait } from '../utils/timeout';
import { createMockProxy } from 'jest-mock-proxy/lib';
import AsyncStore from '../storage/AsyncStore';
import { createStackNavigator } from '@react-navigation/stack';
import { User } from '../typings/user';
import { RootScreens } from '../screens';
import HomeScreen from './HomeScreen';

const mockStore = createMockProxy<AsyncStore>();

const Stack = createStackNavigator();

const wrapComponent = (cmp: ReactNode) => (
  <Provider store={store}>
    <AsyncStorageProvider defaultStorage={mockStore}>
      <ThemeProvider>
        <NavigationContainer>
          <Stack.Navigator initialRouteName="test">
            <Stack.Screen name="test">{() => cmp}</Stack.Screen>
            <Stack.Screen name={RootScreens.Home} component={HomeScreen} />
          </Stack.Navigator>
        </NavigationContainer>
      </ThemeProvider>
    </AsyncStorageProvider>
  </Provider>
);

describe('<SplashScreen />', () => {
  beforeEach(() => {
    mockStore.mockClear();
  });

  it('should render without crashing', () => {
    const cmp = render(wrapComponent(<SplashScreen />));
    expect(cmp.toJSON()).toMatchSnapshot();
  });

  it('should show spinner on loading', async () => {
    mockStore.get.mockImplementation(async () => {
      await wait(9000);

      return null;
    });

    const cmp = render(wrapComponent(<SplashScreen />));

    await act(async () => {
      await wait(1100);
    });

    const spinner = cmp.getByTestId('splashSpinner');
    expect(spinner).toBeDefined();
  });

  it('should show "Get Started" btn if user does not exist', async () => {
    const cmp = render(wrapComponent(<SplashScreen />));

    await act(async () => {
      await wait(1100);
    });

    const btn = cmp.getByTestId('getStarted');
    expect(btn).toBeDefined();
  });

  it('should redirect to HomeScreen if user is already created', async () => {
    const user: User = {
      name: 'Test',
      createdAt: new Date(),
      hasBeenWelcomed: false,
    };

    mockStore.get.mockResolvedValue(user);

    const cmp = render(wrapComponent(<SplashScreen />));

    await act(async () => {
      await wait(1100);
    });

    const homeScreen = cmp.findByTestId('homeScreen');
    expect(homeScreen).toBeDefined();
  });
});
