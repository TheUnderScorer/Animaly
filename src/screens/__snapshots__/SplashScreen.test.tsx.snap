// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SplashScreen /> should render without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaProvider
      onInsetsChange={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onOpen={[Function]}
            onTransitionStart={[Function]}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 64,
                }
              }
            >
              <View
                enabled={false}
                hitSlop={
                  Object {
                    "right": -700,
                  }
                }
                maxDeltaY={20}
                minOffsetX={5}
                onHandlerStateChange={[Function]}
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "transform": Array [
                        Object {
                          "translateX": 0,
                        },
                        Object {
                          "translateX": 0,
                        },
                      ],
                    }
                  }
                >
                  <View
                    pointerEvents="none"
                    style={
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "shadowColor": "#000",
                        "shadowOffset": Object {
                          "height": 1,
                          "width": -1,
                        },
                        "shadowOpacity": 0.3,
                        "shadowRadius": 5,
                        "top": 0,
                        "width": 3,
                      }
                    }
                  />
                  <View
                    style={
                      Array [
                        Object {
                          "flex": 1,
                          "overflow": "hidden",
                        },
                        Array [
                          Object {
                            "backgroundColor": "rgb(242, 242, 242)",
                          },
                          undefined,
                        ],
                      ]
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                          "flexDirection": "column-reverse",
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "justifyContent": "center",
                            }
                          }
                          transition={
                            <Transition.Sequence>
                              <Transition.Out
                                type="fade"
                              />
                              <Transition.In
                                delayMs={500}
                                type="fade"
                              />
                              <Transition.Change
                                interpolation="easeInOut"
                              />
                            </Transition.Sequence>
                          }
                        >
                          <View
                            appearance="default"
                            level="4"
                            style={
                              Array [
                                Object {
                                  "backgroundColor": "#E4E9F2",
                                },
                                Object {
                                  "alignItems": "center",
                                  "height": "100%",
                                  "justifyContent": "center",
                                  "width": "100%",
                                },
                              ]
                            }
                          >
                            <Image
                              source={
                                Object {
                                  "testUri": "../../../src/assets/paws.jpg",
                                }
                              }
                              style={
                                Object {
                                  "height": "100%",
                                  "position": "absolute",
                                  "top": 0,
                                  "width": "100%",
                                  "zIndex": 0,
                                }
                              }
                            />
                            <View
                              style={
                                Array [
                                  Object {
                                    "backgroundColor": "rgba(0, 0, 0, 0.3)",
                                    "height": "100%",
                                    "position": "absolute",
                                    "width": "100%",
                                    "zIndex": 0,
                                  },
                                ]
                              }
                            />
                            <Text
                              style={
                                Object {
                                  "bottom": 20,
                                  "color": "#fff",
                                  "left": 10,
                                  "position": "absolute",
                                }
                              }
                            >
                              Designed by
                               
                              <Text
                                appearance="default"
                                category="p1"
                                onPress={[Function]}
                                style={
                                  Array [
                                    Object {
                                      "color": "#222B45",
                                      "fontFamily": "System",
                                      "fontSize": 15,
                                      "fontWeight": "400",
                                    },
                                    Array [
                                      Object {
                                        "color": "blue",
                                        "textDecorationColor": "black",
                                        "textDecorationLine": "underline",
                                        "textDecorationStyle": "solid",
                                      },
                                    ],
                                  ]
                                }
                                variant="dark"
                              >
                                pch-vector
                              </Text>
                            </Text>
                            <View>
                              <View
                                style={
                                  Object {
                                    "marginBottom": 40,
                                  }
                                }
                              >
                                <View
                                  style={
                                    Array [
                                      Object {
                                        "flexDirection": "row",
                                      },
                                      Object {
                                        "alignItems": "center",
                                        "justifyContent": "center",
                                      },
                                    ]
                                  }
                                >
                                  <Text
                                    appearance="default"
                                    category="h1"
                                    style={
                                      Array [
                                        Object {
                                          "color": "#222B45",
                                          "fontFamily": "System",
                                          "fontSize": 36,
                                          "fontWeight": "800",
                                        },
                                        Object {
                                          "color": "#fff",
                                        },
                                      ]
                                    }
                                  >
                                    Animaly
                                  </Text>
                                  <Text
                                    allowFontScaling={false}
                                    style={
                                      Array [
                                        Object {
                                          "color": undefined,
                                          "fontSize": 40,
                                        },
                                        Object {
                                          "color": "#fff",
                                        },
                                        Object {
                                          "fontFamily": "Material Design Icons",
                                          "fontStyle": "normal",
                                          "fontWeight": "normal",
                                        },
                                        Object {},
                                      ]
                                    }
                                  >
                                    暑
                                  </Text>
                                </View>
                                <View
                                  style={
                                    Array [
                                      Object {
                                        "flexDirection": "row",
                                      },
                                      Object {
                                        "alignItems": "center",
                                        "justifyContent": "center",
                                      },
                                      Object {
                                        "marginTop": 5,
                                      },
                                    ]
                                  }
                                >
                                  <Text
                                    appearance="default"
                                    category="s2"
                                    style={
                                      Array [
                                        Object {
                                          "color": "#222B45",
                                          "fontFamily": "System",
                                          "fontSize": 13,
                                          "fontWeight": "600",
                                        },
                                        Object {
                                          "color": "#fff",
                                        },
                                      ]
                                    }
                                  >
                                    Your everyday companion.
                                  </Text>
                                </View>
                              </View>
                              <View
                                style={
                                  Array [
                                    Object {
                                      "flexDirection": "row",
                                    },
                                    Object {
                                      "alignItems": "center",
                                      "justifyContent": "center",
                                      "width": 300,
                                    },
                                  ]
                                }
                              />
                            </View>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        pointerEvents="box-none"
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(224, 224, 224)",
                  "flex": 1,
                  "shadowColor": "rgb(224, 224, 224)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            pointerEvents="box-none"
            style={
              Object {
                "height": 64,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 20,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <View
                pointerEvents="box-none"
                style={
                  Object {
                    "marginHorizontal": 16,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  test
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaProvider>
  </View>
</View>
`;
